src/handlers.rs

```diff
@@ -261,8 +388,200 @@ fn serialize_initgroups(groups: Vec<Gid>) -> Result<Vec<u8>> {
+            + buf_aliases.len()
+            + buf_aliases_size.len();
+
+        let mut buf = Vec::with_capacity(total_len);
+
+        // add header
+        buf.extend_from_slice(header.as_slice());
+
+        // add hostname
+        buf.extend_from_slice(&hostname_c_string_bytes);
+
+        // add aliases sizes
+        buf.extend_from_slice(buf_aliases_size.as_slice());
+
+        // add serialized addresses from buf_addrs
+        buf.extend_from_slice(buf_addrs.as_slice());
+
+        // add aliases
+        buf.extend_from_slice(buf_aliases.as_slice());
+
+        debug_assert_eq!(buf.len(), total_len);
+
+        buf
+    } else {
+        let error_header = protocol::HstResponseHeader {
+            version: protocol::VERSION,
+            found: 0,
+            h_name_len: 0,
+            h_aliases_cnt: 0,
+            h_addrtype: -1,
+            h_length: -1,
+            h_addr_list_cnt: 0,
+            error: hostent.herrno,
+        };
+        Vec::from(error_header.as_slice())
+    };
+
+    Ok(buf)
+}
+
+/// Serialize a [RequestType::GETAI] response to the wire.
+///
+/// This wire format has been implemented by reading the `addhstaiX`
+/// function living in the `nscd/aicache.c` glibc file. We copy the
+/// exact same behaviour, aside from the caching part.
+///
+/// The wire getaddrinfo call result is serialized like this:
+///
+/// 1. version: int32. Hardcoded to 2.
+/// 2. found: int32. 1 if we have a result, 0 if we don't.
+/// 3. naddrs: int32. Number of IPv4/6 adresses we're about to write.
+/// 4. addrslen: int32. Total length of the IPv4/6 adresses we're
+///    about to write.
+/// 5. canonlen: int32. Total length of the null-terminated canonical
+///    name string.
+/// 6. error: int32. Error code. Always 0 in the current nscd
+///           implementation.
+/// 7. addrs: \[BE-encoded IPv4/IPv6\]. We sequentially write the
+///    IPv4 and IPv6 bytes using a big endian encoding. There's no
+///    padding, an IPv4 will be 4 bytes wide, an IPv6 16 bytes wide.
+/// 8. addr_family: \[uint8\]. This array mirrors the addrs array. Each
+///    addr element will be mirrored in this array, except we'll write
+///    the associated IP addr family number. AF_INET for an IPv4,
+///    AF_INET6 for a v6.
+/// 9. canon_name: Canonical name of the host. Null-terminated string.
+fn serialize_address_info(resp: AiResponse) -> Result<Vec<u8>> {
+    let mut b_families: Vec<u8> = Vec::with_capacity(2);
+    let mut b_addrs: Vec<u8> = Vec::with_capacity(2);
+    for addr in &resp.addrs {
+        match addr {
+            IpAddr::V4(ip) => {
+                b_families.push(AddressFamily::Inet as u8);
+                for octet in ip.octets() {
+                    b_addrs.push(octet);
+                }
+            }
+            IpAddr::V6(ip) => {
+                b_families.push(AddressFamily::Inet6 as u8);
+                for segment in ip.segments() {
+                    for byte in u16::to_be_bytes(segment) {
```

Comment:
any reason this doesn't use `octets`?

https://doc.rust-lang.org/stable/std/net/struct.Ipv6Addr.html#method.octets

Reply:
Yes, this shoudl work. Updated.